File: Java/src/main/java/QuestionBank/Solution101_200/Solution106.java
package QuestionBank.Solution101_200;

import tool.TreeNode;

import java.util.HashMap;

/**
 * @author windows
 */
public class Solution106 {
    HashMap<Integer, Integer> map = new HashMap<>();
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        int n = inorder.length;
        for (int i = 0; i < n; i++) {
            map.put(inorder[i], i);
        }
        return process(inorder, postorder, 0, n - 1, 0, n - 1);
    }

    private TreeNode process(int[] inorder, int[] postorder, int inorderBegin, int inorderEnd, int postorderBegin, int postorderEnd){
        // [9,3,15,20,7]    [9,15,7,20,3]
        if (postorderBegin > postorderEnd){
            return null;
        }
        TreeNode root = new TreeNode(postorder[postorderEnd]);
        int rootIndex = map.get(postorder[postorderEnd]);
        int leftSize = rootIndex - inorderBegin;
        int rightSize = inorderEnd - rootIndex;
        root.left = process(inorder, postorder, inorderBegin, rootIndex - 1, postorderBegin,  postorderEnd + leftSize);
        root.right = process(inorder, postorder, rootIndex + 1, inorderEnd, postorderEnd - 1 - rightSize, postorderEnd - 1);
        return root;
    }


    public static void main(String[] args) {
        Solution106 solution106 = new Solution106();
        System.out.println(solution106.buildTree(new int[]{9, 3, 15, 20, 7}, new int[]{9, 15, 7, 20, 3}));
    }
}

File: C++/src/Solution101_200/Solution106.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution106.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

int index(vector<int>& inorder, int target){
  int n = inorder.size();
  for (int i = 0; i < n; ++i) {
    if (inorder[i] == target){
      return i;
    }
  }
  return -1;
}

TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
  if (inorder.size() == 0){
    return nullptr;
  }
  int left_size = index(inorder, postorder[postorder.size() - 1]);

  vector<int> left_inorder(inorder.begin(), inorder.begin() + left_size);
  vector<int> right_inorder(inorder.begin() + left_size + 1, inorder.end());

  vector<int> left_postorder(postorder.begin(), postorder.begin() + left_size);
  vector<int> right_postorder(postorder.begin() + left_size, postorder.end() - 1);

  TreeNode* left = buildTree(left_inorder, left_postorder);
  TreeNode* right = buildTree(right_inorder,right_postorder);

  TreeNode* root = new TreeNode(postorder[postorder.size() - 1], left, right);
  return root;
}
